react/features/vad-reporter/TrackVADEmitter.js

```diff
@@ -0,0 +1,281 @@
+     */
+    _publishVAD: (score: VADScore) => void;
+
+    /**
+     * State flag, check if the instance was destroyed
+     */
+    _destroyed: boolean = false;
+
+    /**
+     * Constructor.
+     *
+     * @param {number} procNodeSampleRate - Sample rate of the proc node.
+     * @param {Function} publishVAD - VAD score callback function.
+     * @param {RnnoiseProcessor} rnnoiseProcessor - Rnnoise adapter that allows us to calculate VAD score
+     * for PCM samples.
+     * @param {Object} jitsiLocalTrack - JitsiLocalTrack corresponding to micDeviceId.
+     */
+    constructor(
+            procNodeSampleRate: number,
+            publishVAD: Function,
+            rnnoiseProcessor: RnnoiseProcessor,
+            jitsiLocalTrack: Object
+    ) {
+        this._procNodeSampleRate = procNodeSampleRate;
+        this._publishVAD = publishVAD;
+        this._rnnoiseProcessor = rnnoiseProcessor;
+        this._jitsiLocalTrack = jitsiLocalTrack;
+        this._micDeviceId = jitsiLocalTrack.getDeviceId();
+        this._bufferResidue = new Float32Array([]);
+        this._audioContext = new AudioContext();
+
+        this._onAudioProcess = this._onAudioProcess.bind(this);
+
+        this._initializeAudioContext();
+        this._connectAudioGraph();
+
+        logger.log(`Constructed VAD emitter for device: ${this._micDeviceId}`);
+    }
+
+    /**
+     * Creates a lib-jitsi-meet track associated with a target microphone.
+     *
+     * @param {string} micDeviceId - Target microphone device id.
+     *
+     * @returns {Promise<jitsiLocalTrack>}
+     */
+    static _getAudioStream(micDeviceId: string) {
+        return JitsiMeetJS.createLocalTracks({
+            devices: [ 'audio' ],
+            micDeviceId
+        }).then(result => {
+            // Because we specify the deviceId there should be a single stream available.
+            if (result[0] === undefined) {
+                throw new Error('Failed to create jitsi local track.');
+            }
+
+            return result[0];
+        });
+    }
+
+    /**
+     * Factory method that sets up all the necessary components for the creation of the TrackVADEmitter.
+     *
+     * @param {string} micDeviceId - Target microphone device id.
+     * @param {number} procNodeSampleRate - Sample rate of the proc node.
+     * @param {Function} publishVAD - VAD score callback function.
+     * @returns {Promise<TrackVADEmitter>} - Promise resolving in a new instance of TrackVADEmitter.
+     */
+    static create(micDeviceId: string, procNodeSampleRate: number, publishVAD: Function) {
+        // Maintain closure variables so we can clean everything up in case the promise chain breaks at any point.
+        let closureRnnoiseProc = null;
+        let closureLocalTrack = null;
+
+        logger.log(`Initializing TrackVADEmitter for device: ${micDeviceId}`);
+
+        return createRnnoiseProcessor()
+            .then(rnnoiseProcessor => {
```

Comment:
It would definitely look and read better. The reason why I didn't use it is because I tried to stick to the general coding practise of the whole project, and I don't remember seeing any async awaits mostly just Promises. But I can definitely change it if acceptable.

Reply:
This was changed to async so I'll mark it as resolved.